### ISSUE TITLE
Define Indexed storage API in ZCF for contract

### ISSUE THREAD

**Issue by rowgraus** [Agoric/agoric-sdk#1917]
**What is the Problem Being Solved?**

**Description of the Design**

**Security Considerations**

**Test Plan**

**Comment by erights** [comment]
@dtribble and I had a good discussion of this over the weekend that we need to recap. I'll let @dtribble speak for his takeaways. Mine is that ZCF itself should do essentially nothing here. The issue is more liveslots and clever storage representations. Will explain.

**Comment by erights** [comment]
IOW, all the interesting work on this one moves to #1918

**Comment by katelynsills** [comment]
@erights - does that change this task (to define useful storage APIs for the contracts, such that Simple Exchange for example isn't building its own data structures)? This task was already imagining all the actual work is done elsewhere, so I wasn't sure. According to the original task, what would actually be done while waiting for the kernel / liveslots-level changes is making Stores/WeakStores in ZCF for the contracts. Then the storage API would remain the same as the implementation changes underneath to use the kernel / liveslots cleverness.

**Comment by erights** [comment]
"Elsewhere" would be liveslots and helpers, not zcf. While waiting for liveslots, helpers-only. Let's discuss. At 2pm meeting maybe, after #1905?

**Comment by katelynsills** [comment]
Sounds good!

**Comment by katelynsills** [comment]
I think the above comments on this issue show there's a disagreement in what this issue is meant to be. My understanding of this task is: zcf should provide an API to contracts to allow them to store offers for later, efficient retrieval. The first implementation of this API will not be efficient at all, but the plan is to make it efficient under the hood without any change in the zcf API. If this isn't the correct interpretation, let's talk it out :)

**Comment by dtribble** [comment]
I agree with @katelynsills summary. Even if the library has no ZCF surface changes, we also need to document how ZCF/ZOE use it to swap out everything about each offer/seat.

**Comment by erights** [comment]
I think this is a liveslots issue rather than a zcf issue. But we can postpone resolving this, yes?

**Comment by katelynsills** [comment]
> I think this is a liveslots issue rather than a zcf issue. But we can postpone resolving this, yes?

I don't think so. We need to plan out Zoe tasks for Beta, of which this is one. And this issue is meant to specifically not be anything at the liveslots level. It's to provide a nice ZCF API for contract developers, independently of the actual efficient improvements or anything at the Swingset/vat level.

**Comment by Tartuffo** [comment]
This is being covered by virtual collections.
